test/test-lib/common.ts

```diff
@@ -6,33 +8,86 @@ type PredicateFunction = () => Resolvable<boolean>;
 export class TimedOutError extends TypedError {}
 
 /**
- * Loops `maxCount` times, calling the `checkFn` predicate. Returning `TRUE` from the
+ * Loops for ~`maxWait` ms or `maxCount` times, calling the `checkFn` predicate. Returning `TRUE` from the
  * predicate will cease the looping. If `FALSE` then we delay by `delayMs` milliseconds.
  *
  * If we haven't seen a `TRUE` when the loop finishes, we throw a {@link TimedOutError}
- *
- * @export
- * @param {number} delayMs
- * @param {number} maxCount
- * @param {PredicateFunction} checkFn
  */
 export async function waitFor({
```

Comment:
I came up with a possible alternative so we can avoid changing current default values. What do you think?
```typescript
export async function waitFor({
    delayMs = 100,
    maxCount = 100,
    maxWait,
    checkFn,
}: {
    delayMs?: number;
    maxCount?: number;
    maxWait?: number;
    checkFn: PredicateFunction;
}) {
    // If maxWait is provided, override delayMs and maxCount values
    // and throw an error when maxWait is exceeded.
    let timeout: NodeJS.Timeout | null = null;
    if (maxWait != null) {
        delayMs = Math.max(maxWait / 10, 10);
        maxCount = Math.ceil(maxWait / delayMs);

        // Set timeout for an error throw in case checkFn() takes too long.
        timeout = setTimeout(() => {
            throw new TimedOutError();
        }, maxWait);
    }

    try {
        for (let i = 1; i <= maxCount; i++) {
            console.log(`⌚  Waiting (${i}/${maxCount})...`);
            await delay(delayMs);

            if (await checkFn()) {
                return;
            }
        }

        throw new TimedOutError();
    } finally {
        if (timeout != null) {
            clearTimeout(timeout);
        }
    }
}
```

Reply:
The throw in that setTimeout will not propagate to the promise change (and probably crash the test runner as unhandled exception). I'm afraid that the `Promise.race` is the only way to achieve exactly this, so let me try to do that...

Reply:
I see what you mean. I came up with another way to write the race to make it a bit more clear (at least to me), but I'll leave it up to you:
```typescript
	// Primary wait loop
	const waitLoop = (async () => {
		for (let i = 1; i <= maxCount; i++) {
			console.log(`⌚  Waiting ${delayMs}ms (${i}/${maxCount})...`);
			await setTimeout(delayMs);

			if (await checkFn()) {
				return;
			}
		}

		throw new TimedOutError();
	})();

	// Timeout logic when maxWait is provided
	const timeoutFn = maxWait == null ? undefined : (async () => {
		await setTimeout(maxWait);
		throw new TimedOutError(`Exceeded maxWait ${maxWait}`);
	})();

	// Await the race between the primary wait loop and the timeoutFn
	await Promise.race([waitLoop, timeoutFn].filter(Boolean));
```